components/places/examples/places-utils.rs

```diff
@@ -156,44 +158,92 @@ fn run_native_export(db: &PlacesDb, filename: String) -> Result<()> {
 }
 
 fn sync(
-    db: &PlacesDb,
-    engine_names: Vec<String>,
+    api: &PlacesApi,
+    mut engine_names: Vec<String>,
     cred_file: String,
     wipe: bool,
     reset: bool,
 ) -> Result<()> {
+    let conn = api.open_connection(ConnectionType::Sync)?;
     let cli_fxa = get_cli_fxa(get_default_fxa_config(), &cred_file)?;
 
-    // markh got a bit ahead of himself here :)
-    // This should move to using PlacesApi.
+    // phew - working with traits is making markh's brain melt!
+    // Note also that PlacesApi::sync() exists and ultimately we should
+    // probably end up using that, but it's not yet ready to handle bookmarks.
     let client_info = Cell::new(None);
-    let stores = if engine_names.is_empty() {
-        vec![HistoryStore::new(db, &client_info)]
+    let stores: Vec<Box<dyn Store>> = if engine_names.is_empty() {
+        vec![
+            Box::new(BookmarksStore::new(&conn, &client_info)),
+            Box::new(HistoryStore::new(&conn, &client_info)),
+        ]
     } else {
-        assert!(engine_names.len() == 1 && engine_names[0] == "history");
-        vec![HistoryStore::new(db, &client_info)]
+        engine_names.sort();
+        engine_names.dedup();
+        engine_names
+            .into_iter()
+            .map(|name| -> Box<dyn Store> {
+                match name.as_str() {
+                    "bookmarks" => Box::new(BookmarksStore::new(&conn, &client_info)),
+                    "history" => Box::new(HistoryStore::new(&conn, &client_info)),
+                    _ => unimplemented!("Can't sync unsupported engine {}", name),
+                }
+            })
+            .collect()
     };
-    let mut sync_ping = telemetry::SyncTelemetryPing::new();
-    for store in stores {
+    for store in &stores {
         if wipe {
             store.wipe()?;
         }
         if reset {
             store.reset()?;
         }
+    }
 
-        log::info!("Syncing {}", store.collection_name());
-        if let Err(e) = store.sync(
-            &cli_fxa.client_init.clone(),
-            &cli_fxa.root_sync_key,
-            &mut sync_ping,
-        ) {
-            log::warn!("Sync failed! {}", e);
-            log::warn!("BT: {:?}", e.backtrace());
-        } else {
-            log::info!("Sync was successful!");
-        }
+    // now the syncs.
+    // XXX - unfortunately, history stores global meta in a `history_global_state`,
+    // but that's a global that should be shared between history and bookmarks.
+    // We should consider changing that key name?
+    // Even more unfortunate, places::storage::get_meta is `pub(crate)`, so we
```

Comment:
We could change that and it wouldn't bother me at all. Probably better than having duplication here, tbh.